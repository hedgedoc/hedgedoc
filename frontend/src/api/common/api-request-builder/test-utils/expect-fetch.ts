/*
 * SPDX-FileCopyrightText: 2022 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import { defaultConfig } from '../../default-config'
import { Mock } from 'ts-mockery'

/**
 * Mock fetch api for tests.
 * Check that the given url and options are present in the request and return the given status code.
 *
 * @param expectedUrl the url that should be requested
 * @param requestStatusCode the status code the mocked request should return
 * @param expectedOptions additional options
 */
export const expectFetch = (
  expectedUrl: string,
  requestStatusCode: number,
  expectedOptions: RequestInit,
  responseBody?: unknown
): void => {
  global.fetch = jest.fn((fetchUrl: RequestInfo | URL, fetchOptions?: RequestInit): Promise<Response> => {
    expect(fetchUrl).toEqual(expectedUrl)
    expect(fetchOptions).toStrictEqual({
      ...defaultConfig,
      body: undefined,
      headers: new Headers(),
      ...expectedOptions
    })
    return Promise.resolve(
      Mock.of<Response>({
        status: requestStatusCode,
        statusText: mapCodeToText(requestStatusCode),
        json: jest.fn(() => (responseBody ? Promise.resolve(responseBody) : Promise.reject()))
      })
    )
  }) as typeof global.fetch
}
const mapCodeToText = (code: number): string => {
  switch (code) {
    case 400:
      return 'bad_request'
    case 401:
      return 'forbidden'
    default:
      return 'unknown_code'
  }
}
