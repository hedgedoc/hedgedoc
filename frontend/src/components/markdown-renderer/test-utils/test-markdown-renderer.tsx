/*
 * SPDX-FileCopyrightText: 2022 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import { StoreProvider } from '../../../redux/store-provider'
import type { MarkdownRendererExtension } from '../extensions/_base-classes/markdown-renderer-extension'
import { MarkdownToReact } from '../markdown-to-react/markdown-to-react'
import React, { useMemo } from 'react'

export interface SimpleMarkdownRendererProps {
  content: string
  extensions: MarkdownRendererExtension[]
}

/**
 * A markdown renderer for tests.
 *
 * @param content The content to be rendered.
 * @param extensions The {@link MarkdownRendererExtension MarkdownExtensions} to use for rendering.
 */
export const TestMarkdownRenderer: React.FC<SimpleMarkdownRendererProps> = ({ content, extensions }) => {
  const lines = useMemo(() => content.split('\n'), [content])
  return (
    <StoreProvider>
      <MarkdownToReact
        markdownContentLines={lines}
        markdownRenderExtensions={extensions}
        newlinesAreBreaks={true}
        allowHtml={false}
      />
    </StoreProvider>
  )
}
