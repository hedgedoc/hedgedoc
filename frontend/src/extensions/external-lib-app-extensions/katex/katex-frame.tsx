/*
 * SPDX-FileCopyrightText: 2022 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import { HtmlToReact } from '../../../components/common/html-to-react/html-to-react'
import { testId } from '../../../utils/test-id'
import KaTeX from 'katex'
import 'katex/dist/katex.min.css'
import React, { useMemo } from 'react'
import { ApplicationErrorAlert } from '../../../components/common/application-error-alert/application-error-alert'

interface KatexFrameProps {
  expression: string
  block?: boolean
}

/**
 * Renders a LaTeX expression.
 *
 * @param expression The mathematical expression to render
 * @param block Defines if the output should be a block or inline.
 */
export const KatexFrame: React.FC<KatexFrameProps> = ({ expression, block = false }) => {
  const dom = useMemo(() => {
    try {
      const katexHtml = KaTeX.renderToString(expression, {
        displayMode: block,
        throwOnError: true
      })
      return (
        <HtmlToReact htmlCode={katexHtml} domPurifyConfig={{ ADD_TAGS: ['semantics', 'annotation'] }}></HtmlToReact>
      )
    } catch (error) {
      return (
        <ApplicationErrorAlert className={block ? '' : 'd-inline-block'}>
          {(error as Error).message}
        </ApplicationErrorAlert>
      )
    }
  }, [block, expression])

  return block ? <p {...testId('katex-block')}>{dom}</p> : <span {...testId('katex-inline')}>{dom}</span>
}

export default KatexFrame
