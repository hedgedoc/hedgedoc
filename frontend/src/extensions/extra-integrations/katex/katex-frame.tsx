/*
 * SPDX-FileCopyrightText: 2022 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import { HtmlToReact } from '../../../components/common/html-to-react/html-to-react'
import { testId } from '../../../utils/test-id'
import KaTeX from 'katex'
import 'katex/dist/katex.min.css'
import React, { useMemo } from 'react'
import { Alert } from 'react-bootstrap'

interface KatexFrameProps {
  expression: string
  block?: boolean
}

/**
 * Renders a LaTeX expression.
 *
 * @param expression The mathematical expression to render
 * @param block Defines if the output should be a block or inline.
 */
export const KatexFrame: React.FC<KatexFrameProps> = ({ expression, block = false }) => {
  const dom = useMemo(() => {
    try {
      const katexHtml = KaTeX.renderToString(expression, {
        displayMode: block,
        throwOnError: true
      })
      return (
        <HtmlToReact htmlCode={katexHtml} domPurifyConfig={{ ADD_TAGS: ['semantics', 'annotation'] }}></HtmlToReact>
      )
    } catch (error) {
      return (
        <Alert className={block ? '' : 'd-inline-block'} variant={'danger'}>
          {(error as Error).message}
        </Alert>
      )
    }
  }, [block, expression])

  return block ? <p {...testId('katex-block')}>{dom}</p> : <span {...testId('katex-inline')}>{dom}</span>
}

export default KatexFrame
