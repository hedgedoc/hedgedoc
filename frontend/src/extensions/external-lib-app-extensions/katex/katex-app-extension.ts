/*
 * SPDX-FileCopyrightText: 2023 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import type { CheatsheetExtension } from '../../../components/cheatsheet/cheatsheet-extension'
import type { MarkdownRendererExtension } from '../../../components/markdown-renderer/extensions/_base-classes/markdown-renderer-extension'
import { AppExtension } from '../../_base-classes/app-extension'
import { KatexMarkdownExtension } from './katex-markdown-extension'

/**
 * Adds support for LaTeX rendering using KaTeX to the markdown rendering.
 *
 * @see https://katex.org/
 */
export class KatexAppExtension extends AppExtension {
  buildMarkdownRendererExtensions(): MarkdownRendererExtension[] {
    return [new KatexMarkdownExtension()]
  }

  buildCheatsheetExtensions(): CheatsheetExtension[] {
    return [{ i18nKey: 'katex', readMoreUrl: new URL('https://katex.org/') }]
  }
}
