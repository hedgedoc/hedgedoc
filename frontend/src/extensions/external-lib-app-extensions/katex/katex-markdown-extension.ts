/*
 * SPDX-FileCopyrightText: 2022 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import { MarkdownRendererExtension } from '../../../components/markdown-renderer/extensions/_base-classes/markdown-renderer-extension'
import type { ComponentReplacer } from '../../../components/markdown-renderer/replace-components/component-replacer'
import { KatexReplacer } from './katex-replacer'
import type MarkdownIt from 'markdown-it'
import mathJax from 'markdown-it-mathjax'

/**
 * Adds support for rendering of LaTeX code using KaTeX.
 *
 * @see https://katex.org/
 */
export class KatexMarkdownExtension extends MarkdownRendererExtension {
  public static readonly tagName = 'app-katex'

  public configureMarkdownIt(markdownIt: MarkdownIt): void {
    mathJax({
      beforeMath: `<${KatexMarkdownExtension.tagName}>`,
      afterMath: `</${KatexMarkdownExtension.tagName}>`,
      beforeInlineMath: `<${KatexMarkdownExtension.tagName} data-inline="true">`,
      afterInlineMath: `</${KatexMarkdownExtension.tagName}>`,
      beforeDisplayMath: `<${KatexMarkdownExtension.tagName}>`,
      afterDisplayMath: `</${KatexMarkdownExtension.tagName}>`
    })(markdownIt)
  }

  public buildReplacers(): ComponentReplacer[] {
    return [new KatexReplacer()]
  }

  public buildTagNameAllowList(): string[] {
    return [KatexMarkdownExtension.tagName]
  }
}
