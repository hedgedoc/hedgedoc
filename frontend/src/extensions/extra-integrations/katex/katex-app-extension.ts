/*
 * SPDX-FileCopyrightText: 2022 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import type { CheatsheetExtension } from '../../../components/editor-page/cheatsheet/cheatsheet-extension'
import type { MarkdownRendererExtension } from '../../../components/markdown-renderer/extensions/base/markdown-renderer-extension'
import { AppExtension } from '../../base/app-extension'
import { KatexMarkdownExtension } from './katex-markdown-extension'

/**
 * Adds support for LaTeX rendering using KaTeX to the markdown rendering.
 *
 * @see https://katex.org/
 */
export class KatexAppExtension extends AppExtension {
  buildMarkdownRendererExtensions(): MarkdownRendererExtension[] {
    return [new KatexMarkdownExtension()]
  }

  buildCheatsheetExtensions(): CheatsheetExtension[] {
    return [{ i18nKey: 'katex', readMoreUrl: new URL('https://katex.org/') }]
  }
}
