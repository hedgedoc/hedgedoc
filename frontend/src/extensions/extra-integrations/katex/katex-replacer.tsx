/*
 * SPDX-FileCopyrightText: 2022 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import type { NodeReplacement } from '../../../components/markdown-renderer/replace-components/component-replacer'
import {
  ComponentReplacer,
  DO_NOT_REPLACE
} from '../../../components/markdown-renderer/replace-components/component-replacer'
import { KatexMarkdownExtension } from './katex-markdown-extension'
import { Optional } from '@mrdrogdrog/optional'
import type { Element } from 'domhandler'
import { isTag } from 'domhandler'
import React from 'react'

const KaTeX = React.lazy(() => import(/* webpackChunkName: "katex" */ './katex-frame'))

/**
 * Detects LaTeX syntax and renders it with KaTeX.
 */
export class KatexReplacer extends ComponentReplacer {
  public replace(node: Element): NodeReplacement {
    return this.extractKatexContent(node)
      .map((latexContent) => {
        const isInline = !!node.attribs?.['data-inline']
        return <KaTeX key={'katex'} block={!isInline} expression={latexContent} />
      })
      .orElse(DO_NOT_REPLACE)
  }

  /**
   * Checks the given node for katex expression tags and extracts the LaTeX code.
   *
   * @param node The node to scan for inline or block tags
   * @return An optional that contains the extracted latex code
   */
  private extractKatexContent(node: Element): Optional<string> {
    return this.isKatexTag(node, true)
      ? Optional.of(ComponentReplacer.extractTextChildContent(node))
      : this.extractKatexBlock(node).map((childNode) => ComponentReplacer.extractTextChildContent(childNode))
  }

  /**
   * Checks if the given node is a KaTeX element.
   *
   * @param node the node to check
   * @param expectedInline defines if the found katex element is expected to be an inline or block element.
   * @return {@link true} if the given node is a katex element.
   */
  private isKatexTag(node: Element, expectedInline: boolean) {
    return (
      node.name === KatexMarkdownExtension.tagName && (node.attribs?.['data-inline'] !== undefined) === expectedInline
    )
  }

  /**
   * Checks if the given node is a KaTeX block.
   *
   * @param node the node to check
   * @return The given node if it is a KaTeX block element, {@link undefined} otherwise.
   */
  private extractKatexBlock(node: Element): Optional<Element> {
    return Optional.of(node)
      .filter((node) => node.name === 'p' && node.children?.length > 0)
      .map((node) =>
        node.children.filter(isTag).find((subNode) => (this.isKatexTag(subNode, false) ? subNode : undefined))
      )
  }
}
