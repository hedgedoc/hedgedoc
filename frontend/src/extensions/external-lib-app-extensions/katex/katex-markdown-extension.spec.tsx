/*
 * SPDX-FileCopyrightText: 2023 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import { TestMarkdownRenderer } from '../../../components/markdown-renderer/test-utils/test-markdown-renderer'
import { KatexMarkdownExtension } from './katex-markdown-extension'
import { render, screen } from '@testing-library/react'
import type { KatexOptions } from 'katex'
import { default as KatexDefault } from 'katex'
import type { PropsWithChildren } from 'react'
import React, { Suspense } from 'react'

jest.mock('katex')
jest.mock('../../../components/common/application-error-alert/application-error-alert', () => ({
  ApplicationErrorAlert: ({ children, ...props }: PropsWithChildren) => (
    <div>
      <h3>This is a mock for ApplicationErrorAlert.</h3>
      Props: <code>{JSON.stringify(props)}</code>
      Children:
      <div>{children}</div>
    </div>
  )
}))

describe('KaTeX markdown extensions', () => {
  afterAll(() => {
    jest.resetAllMocks()
    jest.resetModules()
  })

  beforeEach(() => {
    jest.spyOn(KatexDefault, 'renderToString').mockImplementation(
      (tex: string, options?: KatexOptions) => `<span>This is a mock for lib katex with this parameters:</span>
<span>
  <span>tex: ${tex}</span>
  <span>block: ${String(options?.displayMode)}</span>
</span>`
    )
  })

  it('renders a valid inline LaTeX expression', async () => {
    const view = render(
      <Suspense fallback={null}>
        <TestMarkdownRenderer extensions={[new KatexMarkdownExtension()]} content={'$\\alpha$'} />
      </Suspense>
    )
    expect(await screen.findByTestId('katex-inline')).toBeInTheDocument()
    expect(view.container).toMatchSnapshot()
  })

  it('renders a valid block LaTeX expression in a single line', async () => {
    const view = render(
      <Suspense fallback={null}>
        <TestMarkdownRenderer extensions={[new KatexMarkdownExtension()]} content={'$$$\\alpha$$$'} />
      </Suspense>
    )
    expect(await screen.findByTestId('katex-block')).toBeInTheDocument()
    expect(view.container).toMatchSnapshot()
  })

  it('renders a valid block LaTeX expression in multi line', async () => {
    const view = render(
      <Suspense fallback={null}>
        <TestMarkdownRenderer extensions={[new KatexMarkdownExtension()]} content={'$$$\n\\alpha\n$$$'} />
      </Suspense>
    )
    expect(await screen.findByTestId('katex-block')).toBeInTheDocument()
    expect(view.container).toMatchSnapshot()
  })

  it('renders an error message for an invalid LaTeX expression', async () => {
    jest.spyOn(KatexDefault, 'renderToString').mockImplementation(() => {
      throw new Error('mocked parseerror')
    })

    const view = render(
      <Suspense fallback={null}>
        <TestMarkdownRenderer extensions={[new KatexMarkdownExtension()]} content={'$\\a$'} />
      </Suspense>
    )
    expect(await screen.findByTestId('katex-inline')).toBeInTheDocument()
    expect(view.container).toMatchSnapshot()
  })
})
