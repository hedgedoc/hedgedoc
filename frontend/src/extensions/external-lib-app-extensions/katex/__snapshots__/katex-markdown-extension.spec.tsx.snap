// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KaTeX markdown extensions renders a valid block LaTeX expression in a single line 1`] = `
<div>
  <p
    data-testid="katex-block"
  >
    <span>
      This is a mock for lib katex with this parameters:
    </span>
    

    <span>
      
  
      <span>
        tex: \\alpha
      </span>
      
  
      <span>
        block: true
      </span>
      

    </span>
  </p>
  

</div>
`;

exports[`KaTeX markdown extensions renders a valid block LaTeX expression in multi line 1`] = `
<div>
  <p
    data-testid="katex-block"
  >
    <span>
      This is a mock for lib katex with this parameters:
    </span>
    

    <span>
      
  
      <span>
        tex: 
\\alpha

      </span>
      
  
      <span>
        block: true
      </span>
      

    </span>
  </p>
  

</div>
`;

exports[`KaTeX markdown extensions renders a valid inline LaTeX expression 1`] = `
<div>
  <p>
    <span
      data-testid="katex-inline"
    >
      <span>
        This is a mock for lib katex with this parameters:
      </span>
      

      <span>
        
  
        <span>
          tex: \\alpha
        </span>
        
  
        <span>
          block: false
        </span>
        

      </span>
    </span>
  </p>
  

</div>
`;

exports[`KaTeX markdown extensions renders an error message for an invalid LaTeX expression 1`] = `
<div>
  <p>
    <span
      data-testid="katex-inline"
    >
      <div
        class="fade d-inline-block alert alert-danger show"
        role="alert"
      >
        mocked parseerror
      </div>
    </span>
  </p>
  

</div>
`;
